MFEM_MAKE_PATH (macro expansion) issues

```
$ make config BUILD_DIR=build-linux CXXFLAGS=-g
for d in general linalg mesh fem config examples examples/sundials examples/petsc examples/pumi miniapps/common miniapps/electromagnetics miniapps/meshing miniapps/performance miniapps/tools 
miniapps/nurbs doc tests/unit; do mkdir -p build-linux/${d}; done
for dir in "" config/ examples/ examples/sundials/ examples/petsc/ examples/pumi/ miniapps/common/ miniapps/electromagnetics/ miniapps/meshing/ miniapps/performance/ miniapps/tools/ miniapps/
nurbs/ doc/ tests/unit/; do \
   printf "# Auto-generated file.\n%s\n%s\n" \
      "MFEM_DIR = /home/jed/src/mfem" \
      "include \$(MFEM_DIR)/${dir}makefile" \
      > build-linux/${dir}GNUmakefile; done
make -C build-linux/config all
make[1]: Entering directory '/home/jed/src/mfem/build-linux/config'
make[1]: Leaving directory '/home/jed/src/mfem/build-linux/config'
cd "build-linux" && ln -sf "/home/jed/src/mfem/data" .
for hdr in mfem.hpp mfem-performance.hpp; do \
   printf "// Auto-generated file.\n%s\n%s\n" \
   "#define MFEM_BUILD_DIR /home/jed/src/mfem/build-linux" \
   "#include \"/home/jed/src/mfem/${hdr}\"" > build-linux/${hdr}; done

Build destination: build-linux [/home/jed/src/mfem/build-linux]

$ make BUILD_DIR=build-linux 
g++   -g  -DMFEM_BUILD_DIR="/home/jed/src/mfem/build-linux" -c general/optparser.cpp -o build-linux/general/optparser.o
In file included from general/optparser.hpp:15,
                 from general/optparser.cpp:12:
general/../config/config.hpp:24:58: fatal error: /home/jed/src/mfem/build-1/config/_config.hpp: No such file or directory
   24 | #include MFEM_MAKE_PATH(MFEM_BUILD_DIR,config/_config.hpp)
      |                                                          ^
compilation terminated.
make: *** [makefile:389: build-linux/general/optparser.o] Error 1

```

The reason for this is that `-std=gnu++14` is the default g++/clang++ dialect, and all the `gnu*` dialects define `linux=1` (and `unix=1`). The above works with `CXXFLAGS='-g -std=c++11'`, but that's a fragile thing to enforce. Of course this problem also exists with paths like `BUILD_DIR=foo-unix-bar` and also if the parent directory contains any such tokens.

This macro expansion with paths is pretty fragile. Could you make it work with `-I` flags instead?

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -395,11 +395,11 @@ endif()
 set_target_properties(mfem PROPERTIES VERSION "${mfem_VERSION}")
 set_target_properties(mfem PROPERTIES SOVERSION "${mfem_VERSION}")
 
-# If building out-of-source, define MFEM_BUILD_DIR to point to the build
-# directory.
+# If building out-of-source, define MFEM_CONFIG_FILE to point to the config file
+# inside the build directory.
 if (NOT ("${PROJECT_SOURCE_DIR}" STREQUAL "${PROJECT_BINARY_DIR}"))
   target_compile_definitions(mfem PRIVATE
-    "MFEM_BUILD_DIR=${PROJECT_BINARY_DIR}")
+    "MFEM_CONFIG_FILE=\"${PROJECT_BINARY_DIR}/config/_config.hpp\"")
 endif()
 
 # Generate configuration file in the build directory: config/_config.hpp.
@@ -415,7 +415,7 @@ if (NOT ("${PROJECT_SOURCE_DIR}" STREQUAL "${PROJECT_BINARY_DIR}"))
       "Writing substitute header --> \"${Header}\"")
     file(WRITE "${PROJECT_BINARY_DIR}/${Header}"
 "// Auto-generated file.
-#define MFEM_BUILD_DIR ${PROJECT_BINARY_DIR}
+#define MFEM_CONFIG_FILE \"${PROJECT_BINARY_DIR}/config/_config.hpp\"
 #include \"${PROJECT_SOURCE_DIR}/${Header}\"
 ")
     # This version will be installed in the top include directory:

--- a/config/config.hpp
+++ b/config/config.hpp
@@ -10,18 +10,15 @@
 // Software Foundation) version 2.1 dated February 1999.
 
 
-// Support out-of-source builds: if MFEM_BUILD_DIR is defined, load the config
-// file MFEM_BUILD_DIR/config/_config.hpp.
+// Support out-of-source builds: if MFEM_CONFIG_FILE is defined, include it.
 //
 // Otherwise, use the local file: _config.hpp.
 
 #ifndef MFEM_CONFIG_HPP
 #define MFEM_CONFIG_HPP
 
-#ifdef MFEM_BUILD_DIR
-#define MFEM_QUOTE(a) #a
-#define MFEM_MAKE_PATH(x,y) MFEM_QUOTE(x/y)
-#include MFEM_MAKE_PATH(MFEM_BUILD_DIR,config/_config.hpp)
+#ifdef MFEM_CONFIG_FILE
+#include MFEM_CONFIG_FILE
 #else
 #include "_config.hpp"
 #endif

--- a/makefile
+++ b/makefile
@@ -128,7 +128,7 @@ BUILD_DIR := $(MFEM_BUILD_DIR)
 BUILD_REAL_DIR := $(abspath $(BUILD_DIR))
 ifneq ($(BUILD_REAL_DIR),$(MFEM_REAL_DIR))
    BUILD_SUBDIRS = $(DIRS) config $(EM_DIRS) doc $(TEST_DIRS)
-   BUILD_DIR_DEF = -DMFEM_BUILD_DIR="$(BUILD_REAL_DIR)"
+   CONFIG_FILE_DEF = -DMFEM_CONFIG_FILE='"$(BUILD_REAL_DIR)/config/_config.hpp"'
    BLD := $(if $(BUILD_REAL_DIR:$(CURDIR)=),$(BUILD_DIR)/,)
    $(if $(word 2,$(BLD)),$(error Spaces in BLD = "$(BLD)" are not supported))
 else
@@ -314,11 +314,11 @@ MFEM_LIBS      ?= $(if $(shared),$(BUILD_RPATH)) -L@MFEM_LIB_DIR@ -lmfem\
 MFEM_LIB_FILE  ?= @MFEM_LIB_DIR@/libmfem.$(if $(shared),$(SO_VER),a)
 MFEM_BUILD_TAG ?= $(shell uname -snm)
 MFEM_PREFIX    ?= $(PREFIX)
-MFEM_INC_DIR   ?= $(if $(BUILD_DIR_DEF),@MFEM_BUILD_DIR@,@MFEM_DIR@)
-MFEM_LIB_DIR   ?= $(if $(BUILD_DIR_DEF),@MFEM_BUILD_DIR@,@MFEM_DIR@)
+MFEM_INC_DIR   ?= $(if $(CONFIG_FILE_DEF),@MFEM_BUILD_DIR@,@MFEM_DIR@)
+MFEM_LIB_DIR   ?= $(if $(CONFIG_FILE_DEF),@MFEM_BUILD_DIR@,@MFEM_DIR@)
 MFEM_TEST_MK   ?= @MFEM_DIR@/config/test.mk
 # Use "\n" (interpreted by sed) to add a newline.
-MFEM_CONFIG_EXTRA ?= $(if $(BUILD_DIR_DEF),MFEM_BUILD_DIR ?= @MFEM_DIR@,)
+MFEM_CONFIG_EXTRA ?= $(if $(CONFIG_FILE_DEF),MFEM_BUILD_DIR ?= @MFEM_DIR@,)
 
 MFEM_SOURCE_DIR  = $(MFEM_REAL_DIR)
 MFEM_INSTALL_DIR = $(abspath $(MFEM_PREFIX))
@@ -382,7 +382,7 @@ lib: $(if $(static),$(BLD)libmfem.a) $(if $(shared),$(BLD)libmfem.$(SO_EXT))
 
 # Flags used for compiling all source files.
 MFEM_BUILD_FLAGS = $(MFEM_PICFLAG) $(MFEM_CPPFLAGS) $(MFEM_CXXFLAGS)\
- $(MFEM_TPLFLAGS) $(BUILD_DIR_DEF)
+ $(MFEM_TPLFLAGS) $(CONFIG_FILE_DEF)
 
 # Rules for compiling all source files.
 $(OBJECT_FILES): $(BLD)%.o: $(SRC)%.cpp $(CONFIG_MK)
@@ -536,7 +536,7 @@ ifeq (,$(and $(findstring B,$(MAKEFLAGS)),$(wildcard $(CONFIG_MK))))
 	$(error )
 endif
 
-config: $(if $(BUILD_DIR_DEF),build-config,local-config)
+config: $(if $(CONFIG_FILE_DEF),build-config,local-config)
 
 .PHONY: local-config
 local-config:
@@ -555,7 +555,7 @@ build-config:
 	cd "$(BUILD_DIR)" && ln -sf "$(MFEM_REAL_DIR)/data" .
 	for hdr in mfem.hpp mfem-performance.hpp; do \
 	   printf "// Auto-generated file.\n%s\n%s\n" \
-	   "#define MFEM_BUILD_DIR $(BUILD_REAL_DIR)" \
+	   "#define MFEM_CONFIG_FILE \"$(BUILD_REAL_DIR)/config/_config.hpp\"" \
 	   "#include \"$(MFEM_REAL_DIR)/$${hdr}\"" > $(BLD)$${hdr}; done
 	@printf "\nBuild destination: $(BUILD_DIR) [$(BUILD_REAL_DIR)]\n\n"
 

